lib/param_parsing/parameters.go

```diff
@@ -56,19 +56,52 @@ func GetParameters() Params {
 	logger = lib.InitLogger(params.LogLevel)
 	var err error
 	// Read configuration files
-	if tomlFileExists(params) {
-		params, err = getParamsTOML(params)
-	} else if tfSwitchFileExists(params) {
+	// TOML from Homedir
+	if tomlFileExists(lib.GetHomeDirectory()) {
+		params, err = getParamsTOML(params, lib.GetHomeDirectory())
+		if err != nil {
+			logger.Fatalf("Failed to obtain settings from TOML config in home directory: %q", err)
+		}
+	}
+
+	// TOML from ChDirPath
+	if tomlFileExists(params.ChDirPath) {
+		params, err = getParamsTOML(params, params.ChDirPath)
+		if err != nil {
+			logger.Fatalf("Failed to obtain settings from TOML config in directory %s: %q", params.ChDirPath, err)
+		}
+	}
+
+	if tfSwitchFileExists(params) {
 		params, err = GetParamsFromTfSwitch(params)
-	} else if terraformVersionFileExists(params) {
+		if err != nil {
+			logger.Fatalf("Failed to obtain settings from tfswitch config: %q", err)
+		}
+	}
+
+	if terraformVersionFileExists(params) {
 		params, err = GetParamsFromTerraformVersion(params)
-	} else if isTerraformModule(params) {
+		if err != nil {
+			logger.Fatalf("Failed to obtain settings from Terraform version: %q", err)
+		}
+	}
+
+	if isTerraformModule(params) {
 		params, err = GetVersionFromVersionsTF(params)
-	} else if terraGruntFileExists(params) {
+		if err != nil {
+			logger.Fatalf("Failed to obtain settings from .tfversions: %q", err)
+		}
+	}
+
+	if terraGruntFileExists(params) {
 		params, err = GetVersionFromTerragrunt(params)
-	} else {
-		params = GetParamsFromEnvironment(params)
+		if err != nil {
+			logger.Fatalf("Error parsing configuration file: %q", err)
+		}
 	}
+
+	params = GetParamsFromEnvironment(params)
+
 	if err != nil {
 		logger.Fatalf("Error parsing configuration file: %q", err)
 	}
```

Comment:
Which `err` this relies upon eventually? Seems like this error handling is sort of redundant, isn't it? 🤔 
If it is not redundant, please use `%v` for errors instead of `%q`.

Reply:
I agree. I was wondering if _any_ of this should actually cause an error.
If I'm writing terraform in a directory and want to get a terraform version initialised, if my terraform is invalid, should it mean that tfswitch errors out completely?

Reply:
I'd support `tfswitch` to fail if any of config file is broken in any way (unreadable because of the the file perms, incorrect parameter or value, so on). As in ignore specific config file only if it doesn't exist.
So we actually could be ignoring "broken" config files to kinda "improve" user experience by trying to guess what they want, though I'm basing off of the point that if the user expects specific config file to provide TF version, we should explicitly fail and let user know they made a typo or any other error and hence they don't get the outcome they're expecting.

Reply:
I would like tfswitch to fail if the configuration file is incorrect. This is because some of these switchings are done by robots or automation, and it's better if tfswitch fails in this case. We don't want to presume what the user wants if tfswitch is being run by a robot or automation.

Reply:
Bringing this back up, from a discussion in #441, we actually ignore any terraform parsing failures and don't even tell the user - do we wish to continue doing this?
Couple of thoughts around this, based on #441 and my own personal experience:
 * I have seen uses where CI/CD pipelines use it and adhere to whatever the user has asked for. If they expect it to happen in terraform and they have a small syntax error (say during CI-part of CI/CD), do we want tfswitch to hang-around waiting for user-input?
 * If there's bad terraform, do we want to fail entirely - maybe there's a new syntax added that we don't support and the parsing fails? (though I guess this would become HCL3 vs HCL2 - but more importantly, are there syntax changes between HCL1 and HCL2 that are not valid and would this stop users of terraform <0.11 from being able to use tfswitch at all?!

Reply:
> do we want tfswitch to hang-around waiting for user-input?

For sure heavy "NO" if this is about CI.
We probably can check whether tty/vty is allocated and only come up with the interactive selection then and explicitly fail otherwise to force CI pipeline to fail too.

> If there's bad terraform, do we want to fail entirely

That's a hard question. I **personally* would prefer `tfswitch` to fail explicitly on bad terraform code (when it fails to load module). While others may prefer `tfswitch` to be less harsh I'd presume. And I'm leaning to think the latter is what ordinary users would prefer most.
Based on our experience from #441 I'd instead like `tfswitch` to log informational messages (hidden by default unless user increases verbosity) for when (and why) it ignored some specific source of TF version (this would allow us to know the reason for #441 w/o conducting "investigation").

Reply:
> log informational messages (hidden by default unless user increases verbosity) for when (and why) it ignored some specific source of TF version

And, as you correctly noticed in the other thread, print explicit log message if the source is discarded because of the incorrect syntax whatsoever.

Reply:
I guess, finally, should this be covered in this PR, or should we re-open #441 and add way more logging around the fact?

Reply:
Though, I will say, hanging CI/CD is _probably_ worse than it erroring saying "invalid terraform - use <version> arg to specify a version".

> We probably can check whether tty/vty is allocated 
I did start to look down this in golang, very briefly before posting and nothing jumped out immediately - I'll take another look

Reply:
> I guess, finally, should this be covered in this PR

I'm not a dev and I prefer less friction, hence I choose this option, though I'd like to relay this decision to you so that's you to decide either to work on this in a standalone PR or in this current one.

> hough, I will say, hanging CI/CD is _probably_ worse than it erroring saying "invalid terraform - use arg to specify a version".

100% agree.

> I did start to look down this in golang, very briefly before posting and nothing jumped out immediately

Probably https://github.com/mattn/go-isatty ?

Reply:
. Probably https://github.com/mattn/go-isatty or https://pkg.go.dev/golang.org/x/term#IsTerminal? (didn't compare them, though we need to ensure that would work on all the supported OS)

ooo - I was _clearly_ googling the wrong thing :facepalm: 

> 100% agree.

:+1: 


> I'm not a dev and I prefer less friction, hence I choose this option, though I'd like to relay this decision to you so that's you to decide either to work on this in a standalone PR or in this current one.

I'm just aware of the amount of changes that go in - if something breaks post-release, do we roll forward or back (and remove a load of other stuff). There's also other bug reports that are pending the PR.
Historically, at least in my small amount of OSS that people actually use :P (so ignore most of it ;) ), I do like smaller things - release fast (release per issue) etc. But I'm open :)  I'm very much aware of @warrensbox bringing us into his project and don't want to dictate things too much (so I'm always up for his suggestions/opinions)

Reply:
If that's a big portion of work, then defo go with standalone PR.

Reply:
Maybe not a big portion of work.. but should be considered and carefully tested and I'd be worried about both of these if it's "tacked on" to push through a PR, if you see what I mean :)

Reply:
On that note, I do fancy a release branch, if we're aiming to release several PRs in one release.
We can get the PRs merged into the branch, so they're all in one place and line up any smaller changes that we can get in, without a big PR to to be "reviewed" (in the literal line-by-line idea), because they've already been reviewed in individual PRs

Reply:
Given existence of any kind of hesitation/worrisome bits/loveless phase of moon 🤪 , defo go with standalone PR =)
I worded that weirdly and effectively I meant if it's the "2-4 lines of code" when I said I prefer this to be part of the current PR 😂

Reply:
> loveless phase of moon
:laughing: :laughing: :laughing: :laughing:

Reply:
I've added a warning when we fail to parse Terraform in b2a5916

example:
```
➜  tmp git:(420-how-to-deal-with-toml-file) ✗ go run ../
16:41:37.128 WARN Error whilst parsing Terraform: "Unclosed configuration block: There is no closing brace for this block before the end of the file. This may be caused by incorrect brace nesting elsewhere in this file."  
Use the arrow keys to navigate: ↓ ↑ → ← 
? Select Terraform version: 
  ▸ 1.8.4 *recent
    1.3.6 *recent

```

Reply:
Have created https://github.com/warrensbox/terraform-switcher/issues/451 for the TTY detection etc.